.upstream-tests/test/cuda/proclaim_return_type.pass.cpp

```diff
@@ -0,0 +1,52 @@
+//===----------------------------------------------------------------------===//
```

Comment:
Even though it is really intended only for `__device__` lambdas, we should make sure it works for other kinds of callables too. 

So I'd add tests that ensure `cuda::proclaim_return_type` work for all of:
- Unannotated lambdas
- `__device__` lambdas
- `__host__ __device__` lambdas
- Named function objects

To make this easier, I'd factor your existing tests into a helper function that takes the callable as a parameter:
```
template <typename UnaryCallable>
void test_proclaim_return_type( UnaryCallable f){
   // tests here
}

struct function_object{
   auto operator()(){...}
};

int main(){
   test_proclaim_return_type([](){...});
   test_proclaim_return_type([]__device__(){...});
   test_proclaim_return_type([] __host__ __device__(){...});
   test_proclaim_return_type(function_object{});
}

Reply:
That's a good point! I used to have a misconception about `__host__ __device__` specifiers so I though that I can't call a `__device__` function from a `__host__ __device__` one. 

I've updated the implementation and tests.